### PR TITLE
Keep type lenses stable

### PR DESCRIPTION
**Pull request by July541**
I got the same effect without using `PositionMapping` compared to instance lenses, but don't know why.

Can anyone explain this?

### CODE REVIEW

**Comment by michaelpj** [comment]
I think you should still need the position mapping otherwise the edit might go in the wrong place? It doesn't look like there's a `usesWithStale` that just gives you _one_ `PositionMapping` for the lot, which feels like an oversight (probably what you want if you're getting multiple stale pieces of info). I guess otherwise you could try to use the `PositionMapping` for the bit you get positions from?

**Comment by July541** [comment]
Introducing `useWithStale` need some refactoring. Taking this opportunity, I'm wondering if it is worth extracting type lens code to a new plugin? 

cc @michaelpj @pepeiborra

**Comment by michaelpj** [comment]
No particular opinion from me, I can't remember if we have any clear policy about where we're going with the `ghcide` plugins.

**Comment by July541** [comment]
Replace `use` with `useWithStale` without more design.

I'll open a new pr to refactor these messed code.

**Comment by July541** [comment]
@michaelpj Would you mind doing a review, I don't want mix bug fixing and refactoring

File: ghcide/src/Development/IDE/Plugin/TypeLenses.hs [hunk 1]
```diff
@@ -99,46 +103,55 @@ properties = emptyProperties
...
+        generateLensForGlobal mp sig@GlobalBindingTypeSig{..} = do
+            range <- toCurrentRange mp =<< srcSpanToRange (gbSrcSpan sig)
+            tedit <- gblBindingTypeSigToEdit sig (Just gblSigsMp)
             let wedit = toWorkSpaceEdit [tedit]
             pure $ generateLens pId range (T.pack gbRendered) wedit
-          gblSigs' = maybe [] (\(GlobalBindingTypeSigsResult x) -> x) gblSigs
-          generateLensFromDiags f =
-            sequence
-              [ pure $ generateLens pId _range title edit
+        generateLensFromDiags mp f =
```
Comment (michaelpj) [inline ghcide/src/Development/IDE/Plugin/TypeLenses.hs:138]: Should this be using a mapping at all? The functions for getting the diagnostics don't give us a mapping, instead you're using `bindingsMp`, is that correct? It does seem like the functions for getting diagnostics should give us a mapping if they can be stale?

File: ghcide/src/Development/IDE/Plugin/TypeLenses.hs [hunk 2]
```diff
@@ -99,46 +103,55 @@ properties = emptyProperties
...
+    (gblSigs@(GlobalBindingTypeSigsResult gblSigs'), gblSigsMp) <-
+      handleMaybeM "Unable to GetGlobalBindingTypeSigs"
+      $ liftIO
+      $ runAction "codeLens.GetGlobalBindingTypeSigs" ideState (useWithStale GetGlobalBindingTypeSigs nfp)
+
+    diag <- liftIO $ atomically $ getDiagnostics ideState
+    hDiag <- liftIO $ atomically $ getHiddenDiagnostics ideState
+
+    let toWorkSpaceEdit tedit = WorkspaceEdit (Just $ Map.singleton uri $ List tedit) Nothing Nothing
+        generateLensForGlobal mp sig@GlobalBindingTypeSig{..} = do
```
Comment (michaelpj) [inline ghcide/src/Development/IDE/Plugin/TypeLenses.hs:133]: This is always called with `globalSigsMp`, and that's in scope at the definition, so maybe just use it directly and don't have the parameter?

File: ghcide/src/Development/IDE/Plugin/TypeLenses.hs [hunk 3]
```diff
@@ -187,19 +212,21 @@ suggestLocalSignature isQuickFix mEnv mTmr mBindings Diagnostic{_message, _range
     , startOfLine <- Position (_line _start) startCharacter
     , beforeLine <- Range startOfLine startOfLine
     , title <- if isQuickFix then "add signature: " <> signature else signature
-    , action <- TextEdit beforeLine $ signature <> "\n" <> T.replicate (fromIntegral startCharacter) " " =
+    , range' <- fromMaybe beforeLine (flip toCurrentRange beforeLine =<< mmp)
```
Comment (michaelpj) [inline ghcide/src/Development/IDE/Plugin/TypeLenses.hs:215]:
```suggestion
    , Just range' <- flip toCurrentRange beforeLine =<< mmp
```
?

File: ghcide/src/Development/IDE/Plugin/TypeLenses.hs [hunk 4]
```diff
@@ -187,19 +212,21 @@ suggestLocalSignature isQuickFix mEnv mTmr mBindings Diagnostic{_message, _range
...
-gblBindingTypeSigToEdit :: GlobalBindingTypeSig -> Maybe TextEdit
-gblBindingTypeSigToEdit GlobalBindingTypeSig{..}
+gblBindingTypeSigToEdit :: GlobalBindingTypeSig -> Maybe PositionMapping -> Maybe TextEdit
+gblBindingTypeSigToEdit GlobalBindingTypeSig{..} mmp
   | Just Range{..} <- srcSpanToRange $ getSrcSpan gbName
     , startOfLine <- Position (_line _start) 0
-    , beforeLine <- Range startOfLine startOfLine =
-    Just $ TextEdit beforeLine $ T.pack gbRendered <> "\n"
+    , beforeLine <- Range startOfLine startOfLine
+    , range' <- fromMaybe beforeLine (flip toCurrentRange beforeLine =<< mmp)
```
Comment (michaelpj) [inline ghcide/src/Development/IDE/Plugin/TypeLenses.hs:228]: ditto

File: ghcide/src/Development/IDE/Plugin/TypeLenses.hs (hunk 3 above)
Comment (michaelpj) [reply ghcide/src/Development/IDE/Plugin/TypeLenses.hs:215]: Or rather, the logic seems slightly off. You want to return `beforeLine` if `mmp` is `Nothing`, but fail the guard if `toCurrentRange` is `Nothing`, I think.

File: ghcide/src/Development/IDE/Plugin/TypeLenses.hs (hunk 1 above)
Comment (michaelpj) [reply ghcide/src/Development/IDE/Plugin/TypeLenses.hs:138]: Oh this is confusing. It takes a _function_ and applies that, and _that_ will have had a position mapping applied to it, which had better be the same as this one :exploding_head:

**Review by michaelpj — APPROVED** [review]
A few nits but generally looks fine.

I generally feel like we should have a `usesWithStale` that returns a _single_ position mapping for all the values, so you can request many things and get a single coherent stale mapping. The current setup probably works but feels risky to me.

File: ghcide/src/Development/IDE/Plugin/TypeLenses.hs (hunk 2 above)
Comment (July541) [reply ghcide/src/Development/IDE/Plugin/TypeLenses.hs:133]: `gbRendered` is the parameter, I marked this explicitly.

File: ghcide/src/Development/IDE/Plugin/TypeLenses.hs (hunk 3 above)
Comment (July541) [reply ghcide/src/Development/IDE/Plugin/TypeLenses.hs:215]: Yes, I want to return `beforeLine` if `mmp` is Nothing or `toCurrentRange` returns Nothing, the guard seems won't fail.🤔

**Comment by July541** [comment]
Stale lens for diagnostics doesn't make sense, because we always have fresh diagnostics even if current module parsed failed( the diagnostic is `parse failed`).

So there is nothing to do if we generate lenses from diagnostics.

File: ghcide/src/Development/IDE/Plugin/TypeLenses.hs (hunk 1 above)
Comment (July541) [reply ghcide/src/Development/IDE/Plugin/TypeLenses.hs]: See https://github.com/haskell/haskell-language-server/pull/3558#issuecomment-1518637683

File: ghcide/src/Development/IDE/Plugin/TypeLenses.hs (hunk 3 above)
Comment (michaelpj) [reply ghcide/src/Development/IDE/Plugin/TypeLenses.hs]: I think we shouldn't return if `toCurrentRange` returns `Nothing`. That means we have a mapping but we can't get it to work. Rather I think the guard should just fail! Same as here: https://github.com/haskell/haskell-language-server/pull/3570#discussion_r1174427561

So it's something like:
```
Just range' <- case mmp of { Nothing -> Just beforeLine; Just mp -> toCurrentRange mp beforeLine; }
```

**Comment by michaelpj** [comment]
> Stale lens for diagnostics doesn't make sense, because we always have fresh diagnostics even if current module parsed failed

But don't we mix together things we got from the diagnostics with things we got from the bindings? So the locations could be all messed up there? I haven't quite understood what's going on there.

**Comment by July541** [comment]
> But don't we mix together things we got from the diagnostics with things we got from the bindings? So the locations could be all messed up there? I haven't quite understood what's going on there.

We don't. The secret is `getDiagnostics` will never fail, which means the lenses from diagnostic will flicker but others from parsed module won't.

We have 3 modes for type lenses, only lenses from `generateLensForGlobal` should keep stable, `suggestLocalSignature` relies on diagnostics. Different modes need different methods to generate lenses, so they won't be mixed.
https://github.com/haskell/haskell-language-server/blob/6640ebf33eeb8a62ccd094e2f47f69106acfd200/ghcide/src/Development/IDE/Plugin/TypeLenses.hs#L135-L140

File: ghcide/src/Development/IDE/Plugin/TypeLenses.hs [hunk 5]
```diff
@@ -99,46 +102,54 @@ properties = emptyProperties
...
+    pure $ List $ case mode of
         Always ->
-          pure (catMaybes $ generateLensForGlobal <$> gblSigs')
-            <> generateLensFromDiags (suggestLocalSignature False env tmr bindings) -- we still need diagnostics for local bindings
-        Exported -> pure $ catMaybes $ generateLensForGlobal <$> filter gbExported gblSigs'
-        Diagnostics -> generateLensFromDiags $ suggestSignature False env gblSigs tmr bindings
-    Nothing -> pure []
+          mapMaybe (generateLensForGlobal gblSigsMp) gblSigs'
+            <> generateLensFromDiags
+                (suggestLocalSignature False (Just env) (Just tmr) (Just bindings)) -- we still need diagnostics for local bindings
```
Comment (michaelpj) [inline ghcide/src/Development/IDE/Plugin/TypeLenses.hs:149]: Here it looks like after generating stuff from the diagnostics, we're going to apply this function to the diagnostics. But this function takes `bindings`, which might need position mapping. Maybe I'm totally wrong here, it just seems suspicious.

Comment (July541) [reply ghcide/src/Development/IDE/Plugin/TypeLenses.hs:149]: `suggestLocalSignature` relies on diagnostic, if diagnostics don't have the local signature warning, the `bindings` is useless, and if diagnostic has, that means we parsed success, and the `bindings` is fresh.
https://github.com/haskell/haskell-language-server/blob/6f5a73507f8d9266a486feaf8695c052362b9b95/ghcide/src/Development/IDE/Plugin/TypeLenses.hs#L171-L176

Comment (michaelpj) [reply ghcide/src/Development/IDE/Plugin/TypeLenses.hs:149]: Okay. Worth a comment!

Comment (July541) [reply ghcide/src/Development/IDE/Plugin/TypeLenses.hs:151]: Done

File: ghcide/src/Development/IDE/Plugin/TypeLenses.hs [hunk 6]
```diff
@@ -212,8 +212,9 @@ gblBindingTypeSigToEdit GlobalBindingTypeSig{..} mmp
   | Just Range{..} <- srcSpanToRange $ getSrcSpan gbName
     , startOfLine <- Position (_line _start) 0
     , beforeLine <- Range startOfLine startOfLine
-    , Just mp <- mmp
-    , Just range <- toCurrentRange mp beforeLine
+    -- If `mmp` is `Nothing`, return the original range, it used by lenses from diagnostic,
+    -- otherwise we apply `toCurrentRange`, and the guard should fail if `toCurrentRange` failed.
+    , Just range <- maybe (Just beforeLine) (flip toCurrentRange beforeLine) mmp
```
Comment (July541) [inline ghcide/src/Development/IDE/Plugin/TypeLenses.hs:217]: I mixed the scene if `mmp` is Nothing, thanks ci halt this 😅

**Pull request merged**